[#66492] Split project overview into tabs "Dashboard" and "Overview"

oliverguenther opened the pull request:
https://community.openproject.org/wp/66492

This PR currently also contains:

- [Split up Overview and Dashboard](https://community.openproject.org/wp/66904)
- [Overview widget for Sub-items](https://community.openproject.org/wp/66493)
- [Rework Homescreen blocks to use `WidgetGridComponent`](https://community.openproject.org/wp/67336)
- [Componentize Homescreen Links](https://community.openproject.org/wp/67346)
- [Componentize Homescreen Announcements](https://community.openproject.org/wp/67347)


----

Related PRs based on this branch:

- [x] #20228
- [ ] #20222
- [x] #20267
- [x] #20350

oliverguenther committed:
Split up overview and dashboard page

oliverguenther committed:
Add description widget

oliverguenther committed:
Add subitems widget and status label

myabc commented:
FYI #20039 has overlapping goals and touches many of the same files.

oliverguenther committed:
Adapt tests

oliverguenther committed:
Prevent duplicate registration now that dashboards is a opce

oliverguenther committed:
Replace help text spec now that it only shows the dashboard

myabc commented on frontend/src/app/features/overview/dashboard.component.ts:
```diff
@@ -29,14 +29,19 @@
+  ],
   providers: GRID_PROVIDERS,
-  standalone: false,
+  standalone: true,
```
💯

myabc commented on frontend/src/app/app.module.ts:
```diff
@@ -305,9 +304,6 @@ export function runBootstrap(appRef:ApplicationRef) {
     OpenprojectDashboardsModule,
 
-    // Overview
-    OpenprojectOverviewModule,
```
👍🏻

myabc commented on modules/overviews/config/routes.rb:
```diff
@@ -5,6 +5,11 @@
         to: "overviews/overviews#show",
         as: :project_overview
+
+    get "projects/:project_id/dashboard",
```
It would be good to add a routing example here: `modules/overviews/spec/routing/overviews_routing_spec.rb` (I had intended to refactor these routes with `resource`/`namespace`/`scope`, but didn't get round to it)

myabc commented on frontend/src/app/shared/components/grids/widgets/widgets.service.ts:
```diff
@@ -1,18 +1,248 @@
+        properties: {
+          name: this.I18n.t('js.grid.widgets.project_status.title'),
+        },
+      },
```
The `WidgetProjectStatusBetaComponent` got missed somehow (commit 322b7b9a)

```suggestion
      },
      {
         identifier: 'project_status_beta',
         component: WidgetProjectStatusBetaComponent,
         title: i18n.t('js.grid.widgets.project_status_beta.title'),
         properties: {
           name: i18n.t('js.grid.widgets.project_status_beta.title'),
         },
      },
```

myabc commented on modules/overviews/app/views/overviews/overviews/dashboard.html.erb:
```diff
@@ -0,0 +1,2 @@
+<%= render partial: "page_header" %>
```
Why a partial and not a custom `PageHeaderComponent`?

myabc commented on modules/overviews/app/components/overviews/overview_grid_component.sass:
```diff
@@ -0,0 +1,12 @@
+  grid-template-areas: "description subitems" "description subitems"
+  grid-template-columns: 1fr 1fr
+  grid-template-rows: auto auto auto
+  gap: 1rem
```
Should this be harmonised with the Dashboard grid? 

`$grid--gap-width: 20px`

myabc reviewed (changes requested):
Generally looks good, but there are a few things to be addressed still:

## bugs

I have no idea why the header is showing in Czech locally?

<img width="645" height="157" alt="Image" src="https://github.com/user-attachments/assets/4ea9425c-84b7-477e-8c52-b1aaa48aa67e" />

## widget components

With this PR we now have two types of Widgets living under `modules/overviews/app/components/overviews/widgets/`

1. those that render their own Border Box and header.
2. those that render in a component_wrapper :"turbo-frame".

The first, static type (Subitems, Description components) are only compatible with the Overview, whereas the second type (Project Status BETA component) can only be shown on the Dashboard.

I'm not sure if it's a hard requirement at this point, but even from a code perspective, it would be good to make Widgets reusable. Perhaps we could make the Border Box + header optional, by passing in a `wrapped` flag?

## visuals

- Although this doesn't need to happen in the PR necessarily, the design of the Overview widgets should be harmonised with the widgets (Border Box header colour, in particular).
- The gap between widgets on the Overview and Dashboard pages should be harmonised.
- Overviews widget headers should be bold IMO.

## code maintenance

Can we delete the separate, dashboards module? It's disabled, except in tests and AFAICT does not offer any additional functionality that isn't available here (and even if did, it's code that is likely to succumb to bitrot pretty quickly)

## testing

 - There are various failing specs. #20033 added quite a few specs, although in most cases you'll it's just a matter of a rename to fix.
 - I've been trying to get in the habit of writing component specs - if there is no complex logic then just a sanity check that the component renders is bette than nothing. Admittedly, it's pain upfront, but it does make refactoring a lot easier down the road. You may be get some mileage from using `modules/overviews/spec/components/overviews/widgets/project_status_component_spec.rb` as a basic.

myabc commented on app/components/projects/status_badge_component.rb:
```diff
@@ -0,0 +1,61 @@
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class Projects::StatusBadgeComponent < ApplicationComponent
```
Do we need this? What's the difference between this component and `Projects::StatusBadgeComponent`? In any case, we're not rendering it anywhere yet.

oliverguenther replied:
We're rendering this in the subitems component. What other component is there? There is the status button component, which renders a button and dropdown, not a label

myabc replied:
Sorry, I missed that.

myabc commented:
There are two code maintenance tickets that are related, but don't really belong here (and were also out-of-scope for #20033):

- #20166 
- #20160

lindenthal commented:
Hi @myabc, Can you support me in getting the pullpreview deployment started? Many thanks, Niels

myabc committed:
Refactor into view components

Co-Authored-By: Oliver Günther <mail@oliverguenther.de>

myabc committed:
Make DashboardComponent standalone

Co-Authored-By: Oliver Günther <mail@oliverguenther.de>

myabc commented:
@lindenthal the deployment should be running again. Let me know if you have any other questions! Danke!

myabc committed:
Fix typo in low permissions page creation feature

myabc committed:
Rename Overview to Dashboard in feature specs

myabc committed:
Restore manage_dashboards permission

Restores permission that was defined in the Dashboards module, replacing
most usages of `manage_overview`.

This partially reverts commit c3bbe97f81fce759166052f4c66a575fad0e2dc7.

myabc committed:
Fix Page Header title

myabc committed:
Implement initial Overviews::PageHeaderComponent

Also adds initial component spec.

myabc committed:
[#67255] Widget Box Component - WIP

myabc committed:
Add initial spec for StatusBadgeComponent

myabc committed:
Add initial spec for DescriptionComponent widget

myabc committed:
Add initial spec for SubitemsComponent widget

myabc commented on app/components/projects/status_badge_component.rb:
```diff
@@ -0,0 +1,61 @@
+    )
+  end
+
+  def render?
+    @status.present?
+  end
```
@oliverguenther this defaults to `Projects::Statuses::NOT_SET`. It will never be `nil`.

oliverguenther replied:
I wonder if we should skip render in this case?

myabc replied:
It might be worth asking the design team their opinion. I guess some customers don't use project statuses and seeing grey "not set" icons is just extra noise.

myabc replied:
We discussed with @psatyal earlier today and agreed that nothing should be rendered if the project status is not set. Fixed.

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
Update Homescreen Blocks

oliverguenther committed:
Introduce PermissionRenamer for renaming permissions

oliverguenther committed:
Add migration to rename manage_overview to manage_dashboards

oliverguenther committed:
Adjust spec for overview ShowComponent

oliverguenther committed:
Fix subitems component

oliverguenther committed:
Restore test_selector

myabc committed:
Merge pull request #20228 from opf/implementation/67255-widget-box-component

[#67255] Implement `WidgetBoxComponent`

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
Fix Rubocop offenses

myabc committed:
[#67346] Componentize Homescreen Links

Implements `Homescreen::LinksComponent`.

myabc committed:
[#67347] Componentize Homescreen Announcements

Implements `Homescreen::AnnouncementComponent`.

myabc committed:
Factor out LinkComponent rendering link items

myabc committed:
Skip rendering Status Badge if status not set

myabc committed:
Also remove top border from nested Border Boxes

myabc committed:
Merge branch 'fix/add-widget-margin-mobile' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
[#67336] Rework Homescreen blocks as Widget Grid

Componentizes blocks (inheriting from `Grid::WidgetComponent`) as well
as implementing `BlocksGridComponent`.

Moves responsibility for data loading from controller to individual
block components.

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

bsatarnejad commented on frontend/src/global_styles/content/_widget_box.sass:
```diff
@@ -119,6 +119,15 @@ $widget-box--enumeration-width: 20px
+    padding-left: 0
+
+    // apply special treatment for nested boxes, e.g. Border Box
+    .Box
```
Is it better to set a specific class name for box instead of using primer class?

myabc replied:
@bsatarnejad Yes. @HDinger and I have discussed this and we're going to remove the backs to make `.widget-box` work with nested Border Box, and instead add better list styling. (I'll probably make these changes as part of #20267, even though strictly speaking, that's not the right ticket).

myabc replied:
I've begun making the changes discussed in #20350 (which to make matters even more complicated, branches off #20267)!

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

myabc committed:
Migrate news-beta template to self-closing tags

  Automated conversion with:

      ng generate @angular/core:self-closing-tag

bsatarnejad commented on modules/overviews/app/controllers/overviews/widgets/project_statuses_controller.rb:
```diff
@@ -36,7 +36,7 @@ def update
     if call.success?
       @project = call.result
-      update_via_turbo_stream(component: Overviews::Widgets::ProjectStatusComponent.new(project: @project, current_user:))
+      update_via_turbo_stream(component: Overviews::Widgets::ProjectStatusComponent.new(@project, current_user:))
```
Isn't current_user removed from the component? https://github.com/opf/openproject/pull/20045/files#diff-cfe3ee19c1b11fa029d726f7543e0e5654a396baf2b912bf7bda6459e08547deL38

myabc replied:
`Grids::WidgetComponent` actually allows a `current_user:` to be passed.

However using `Dry::Initializer` (and its meta-programming) rather than an explicit `initialize` method, so I appreciate that that's probably not clear!

https://dry-rb.org/gems/dry-initializer/3.0/params-and-options/

myabc replied:
(I'm still trying to figure out the best way of adding Yardocs for meta-programmed initializers)

myabc commented on modules/grids/spec/components/grids/widget_box_component_spec.rb:
```diff
@@ -0,0 +1,44 @@
+  subject(:rendered) { render_component(title: "Cool Widget") }
+
+  it "renders with default inset styles" do
+    puts rendered
```
Whoops! - i copied and pasted the `InsetBox` spec and forgot to remove this line!

myabc replied:
Fixed on this branch: [e080187](https://github.com/opf/openproject/pull/20045/commits/e080187746fbdedd696cbe9e8d6d7635aeeea072)

myabc committed:
Simplify (YAGNI): inline WidgetedComponent mixin

Having one approach to building widgets (i.e. subclassing
`Grids::WidgetComponent`) should suffice for the time-being.

myabc committed:
Skip rendering Widget Box Header if title is blank

myabc committed:
Begin documenting WidgetComponent

myabc committed:
Remove debug output from WidgetBoxComponent spec

myabc commented on modules/overviews/app/views/overviews/widgets/project_statuses/show.html.erb:
```diff
@@ -27,4 +27,4 @@ See COPYRIGHT and LICENSE files for more details.
 ++#%>
 
-<%= render Overviews::Widgets::ProjectStatusComponent.new(project: @project, current_user:) %>
+<%= render Overviews::Widgets::ProjectStatusComponent.new(@project, current_user:) %>
```
We're inconsistent throughout the codebase. Will ignore for now.

HDinger committed:
Fix news and projects widget

HDinger committed:
Use Octicons for homescreen links

HDinger committed:
Use Banner component for Announcements

HDinger committed:
Make Upsell widget full width

HDinger committed:
Merge pull request #20267 from opf/code-maintenance/67336-homescreen-blocks-widget-grid-component

[#67336] Rework Homescreen blocks using `WidgetGridComponent`

myabc committed:
[#67525] API Improvements to `WidgetComponent` and related components (#20350)

* Rework WidgetComponent to render its own WidgetBox

Adds `rows` slots to `WidgetBoxComponent` with the goal of offering a
similar API to `BorderBox` (albeit subset) and obviating the need for
brittle styling overrides.

Also moves responsibility for rendering `turbo-frame` to the Widget Box.

* Remove suffix from nested Header, Body classes

The `Component` suffix is useful in avoiding collisions with other
objects, especially Rails models, but one `Component` in a
full-qualified class name suffices.

* Fix failing test and pass header id from the widget box component

* Create a new component for showing Rows in the body

* Add component_wrapper to be able to act as streamable

---------

Co-authored-by: Behrokh Satarnejad <b.satarnejad@openproject.com>
Co-authored-by: Henriette Darge <h.darge@openproject.com>

HDinger committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard

HDinger committed:
Move News widget to Overiew module and let it inherit from WidgetComponent

HDinger committed:
Move news widget code to overviews module

HDinger committed:
Re-add global news widget route

HDinger committed:
Merge pull request #20387 from opf/housekeeping/migrate-meetings-to-new-structure

Show news widget on new overview page

bsatarnejad committed:
replace beta news widget with news widget component (#20392)

* replace beta news widget with news widget component

* remove feature flag for beta component

* Add a migration to remove the beta_news component and remove beta component feature flag

* change migration file name

HDinger committed:
Migrate angular subitems widget to rails based implementation

bsatarnejad commented on modules/overviews/spec/routing/widget_routing_spec.rb:
```diff
@@ -120,4 +120,26 @@
+
+  describe "subitems routing" do
+    describe "GET #show" do
+      context "with project" do
```
Can we remove context "with project"?

myabc committed:
Merge branch 'dev' into implementation/66904-split-up-overview-and-dashboard
